swap/peer.go

```diff
@@ -137,3 +120,94 @@ func (sp *Peer) handleErrorMsg(ctx context.Context, msg interface{}) error {
 	// maybe balance disagreement
 	return nil
 }
+
+// processAndVerifyCheque verifies the cheque and compares it with the last received cheque
+// if the cheque is valid it will also be saved as the new last cheque
+func (sp *Peer) processAndVerifyCheque(cheque *Cheque) error {
+	if err := sp.verifyChequeProperties(cheque); err != nil {
+		return err
+	}
+
+	lastCheque := sp.loadLastReceivedCheque()
+
+	// TODO: there should probably be a lock here?
+	expectedAmount, err := sp.swap.oracle.GetPrice(cheque.Honey)
+	if err != nil {
+		return err
+	}
+
+	if err := sp.verifyChequeAgainstLast(cheque, lastCheque, expectedAmount); err != nil {
+		return err
+	}
+
+	if err := sp.saveLastReceivedCheque(cheque); err != nil {
+		log.Error("error while saving last received cheque", "peer", sp.ID().String(), "err", err.Error())
+		// TODO: what do we do here?
+	}
+
+	return nil
+}
+
+// verifyChequeProperties verifies the signautre and if the cheque fields are appropiate for this peer
+// it does not verify anything that requires knowing the previous cheque
+func (sp *Peer) verifyChequeProperties(cheque *Cheque) error {
+	if cheque.Contract != sp.contractAddress {
+		return fmt.Errorf("wrong cheque parameters: expected contract: %x, was: %x", sp.contractAddress, cheque.Contract)
+	}
+
+	// the beneficiary is the owner of the counterparty swap contract
+	if err := sp.swap.verifyChequeSig(cheque, sp.beneficiary); err != nil {
+		log.Error("error invalid cheque", "from", sp.ID().String(), "err", err.Error())
```

Comment:
why do we do a `log` here and in the other cases an `fmt.Errorf`?

Reply:
In the other cases we create an error to return. Here we already get an error from `verifyChequeSig` which we return unmodified (hence no `fmt.Errorf`). The log was just here for logging purposes, I moved it to `processAndVerifyCheque` now so that there is one single log call for all kinds of errors.